test/randtilegen.c

```diff
@@ -0,0 +1,264 @@
+	struct {
+		unsigned char palette;
+		unsigned char nbColors;
+	} attributes[NB_TILES];
+	uint8_t tileData[NB_TILES][8][8];
+	// These two are in tiles, not pixels, and in range [3; 10], hence `NB_TILES` above
+	// Both width and height are 4-bit values, so nbTiles is 8-bit (OK!)
+	uint8_t const width = getRandomBits(3) + 3, height = getRandomBits(3) + 3,
+	              nbTiles = width * height;
+
+	for (uint8_t p = 0; p < nbTiles; p++) {
+		uint8_t pal;
+		do {
+			pal = getRandomBits(5);
+		} while (pal == 0 || (pal > 29));
+		attributes[p].palette = 2 * pal + getRandomBits(1);
+		// Population count (nb of bits set), the simple way
+		static uint8_t const popcount[] = {1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4,
+		                                   1, 2, 2, 3, 2, 3, 3, 4, 2, 3, 3, 4, 3, 4};
+		attributes[p].nbColors = popcount[pal - 1];
+		// Handle single-color tiles the simple way, without trying to pull more random bits
+		if (attributes[p].nbColors < 2) {
+			memset(tileData[p], 0, sizeof(tileData[p]));
+			continue;
+		}
+		uint8_t index, total;
+		for (index = 0, total = 0; index < p; index++) {
+			if (attributes[index].nbColors == attributes[p].nbColors) {
+				total++;
+			}
+		}
+		// index == p at exit
+		if (total) {
+			index = getRandomBits(8);
+			if (index < total) {
+				total = index + 1;
+				for (index = 0; total; index++) {
+					if (attributes[index].nbColors == attributes[p].nbColors) {
+						total--;
+					}
+					if (!total) {
+						index--;
+					}
+				}
+			} else {
+				index = p;
+			}
+		}
+		if (index != p) {
+			unsigned rotation = getRandomBits(2);
+			for (uint8_t y = 0; y < 8; y++) {
+				for (uint8_t x = 0; x < 8; x++) {
+					tileData[p][y][x] =
+					    tileData[index][y ^ ((rotation & 2) ? 7 : 0)][x ^ ((rotation & 1) ? 7 : 0)];
+				}
+			}
+		} else {
+			switch (attributes[p].nbColors) {
+			case 2:
+				for (uint8_t y = 0; y < 8; y++)
+					for (uint8_t x = 0; x < 8; x++)
+						tileData[p][y][x] = getRandomBits(1);
+				break;
+			case 4:
+				for (uint8_t y = 0; y < 8; y++)
+					for (uint8_t x = 0; x < 8; x++)
+						tileData[p][y][x] = getRandomBits(2);
+				break;
+			default:
+				for (uint8_t y = 0; y < 8; y++)
+					for (uint8_t x = 0; x < 8; x++) {
+						do {
+							index = getRandomBits(2);
+						} while (index == 3);
+						tileData[p][y][x] = index;
+					}
+			}
+		}
+	}
+
+	uint16_t colors[10];
+	for (uint8_t p = 0; p < 10; p++) {
+		colors[p] = getRandomBits(15);
+	}
+	// Randomly make color #0 of all palettes transparent
+	if (!getRandomBits(2)) {
+		colors[0] |= 0x8000;
+		colors[5] |= 0x8000;
+	}
+
+	uint16_t palettes[60][4];
+	for (uint8_t p = 0; p < 60; p++) {
+		const uint16_t *subpal = colors;
+		if (p & 1) {
+			subpal += 5;
+		}
+		uint8_t total = 0;
+		for (uint8_t index = 0; index < 5; index++) {
+			if (p & (2 << index)) {
+				palettes[p][total++] = subpal[index];
+			}
+		}
+	}
+
+	png_set_IHDR(png, pngInfo, width * 8, height * 8, 8, PNG_COLOR_TYPE_RGB_ALPHA,
```

Comment:
Writing the generated image would be fine as its own function. The raw rgba image buffer could either be a static variable or malloc'ed.

Reply:
Arguably, yeah. I kept the original's structure while I was porting, to try and avoid breakage.